### ISSUE TITLE
`strpos` with a constant first arg

### ISSUE THREAD

**Issue by staabm** [vimeo/psalm#4070]
similar to https://github.com/vimeo/psalm/issues/3877 I guess psalm should emit a error, when someone uses `strpos` with a string literal as a first argument.

I can't think of a good use-case where I want the $haystack to be a string literal.

in contrast when the $haystack and $needle params are mixed up, it can lead to bugs, see e.g.
https://github.com/joomla/joomla-cms/pull/30489
https://github.com/composer/composer/pull/9154
https://github.com/justinrainbow/json-schema/pull/641

https://psalm.dev/r/22d89c087e -> the `strpos()` call should emit an error

**Comment by staabm** [comment]
additionally to my above request, I would expect psalm to error, because strpos() can never return `-1`

**Comment by weirdan** [comment]
@psalm-github-bot, please fetch this: https://psalm.dev/r/22d89c087e

**Comment by muglug** [comment]
Fixed in 422271b

**Comment by staabm** [comment]
Thanks!

**Comment by staabm** [comment]
Hmm wondering why I dont get a error on https://psalm.dev/r/e0c5b3e258 after you fixed this issue?

**Comment by staabm** [comment]
@muglug could you re-open the issue, as it seems not to be fixed atm

**Comment by weirdan** [comment]
> I can't think of a good use-case where I want the $haystack to be a string literal.

This does cause false positives in the code that uses strings as lookup tables (we have quite a few of those for generating barcodes). Can we make the issue more specific, so that it could be selectively suppressed?

**Comment by muglug** [comment]
> Can we make the issue more specific, so that it could be selectively suppressed?

Sure, I'll do that

**Comment by staabm** [comment]
> This does cause false positives in the code that uses strings as lookup tables (we have quite a few of those for generating barcodes)

@weirdan Im curios.. could younshow a example for what you are talking about?

**Comment by staabm** [comment]
@muglug its working now, thx for fixing

One question!

https://psalm.dev/r/22d89c087e

Outputs

> ERROR: DocblockTypeContradiction - 8:7 - int can never contain false|int|int

Shouldn‘t it be

> ERROR: DocblockTypeContradiction - 8:7 - int can never contain positive-int|0|false

?

**Comment by weirdan** [comment]
> Im curios.. could younshow a example for what you are talking about?

Sure. E.g. [Codabar](https://en.wikipedia.org/wiki/Codabar) encoding may look like this: https://psalm.dev/r/ffe1e64c0f

**Comment by muglug** [comment]
@weirdan I think I can add an explicit check for strings with sequential characters, saw a similar false-positive in Vimeo's codebase

**Comment by alphp** [comment]
> I can't think of a good use-case where I want the $haystack to be a string literal.
> 

So how do you code this?
```php
return false === stripos('KLMPQRS', $tinParts[1]) ? $checksum1 : $checksum2;
```
